### DISCUSSION TITLE
How would you handle translations?

### DISCUSSION THREAD

**Discussion by darrylyoung** [prisma/prisma#6084] [Q&A]
This is more of a conceptual question than something directly related to Prisma but I'm hoping some people here will have some thoughts on this.

### Context

I'm using Prisma with Apollo Server to create the GraphQL API used by a React Native app. Products in the database have things like `description`, `notes`, etc. Right now, those things are simply stored as English strings, or lists of strings in the case of `notes`.

### Problem

I've recently been thinking about translations. My app has the ability to provide translations for in-app strings but everything from the database, at least right now, will be in English. I'm wondering what'd be a smart way to include translations for these things straight from the database. Just thinking out loud, maybe I could store things like `description` as `JSON` (I'm using PostgreSQL) instead of a string and then structure that by `"en"`, `"de"`, etc. Then, the user's locale could be passed via query parameter.

Anyway, I'm kind of new to all this and learning as I go so I'd be happy to hear some suggestions.

Thanks, everyone! Have a great day.

**Comment by ryands17** [comment]
@darrylyoung 👋 

Yes one way of doing this would be storing values as a key-pair in `JSON` as `language => text` in the column. In this case, you would need to fetch it in the same way by passing the required language as a key for all the fields. Prisma currently doesn't support [JSON specific filtering](https://github.com/prisma/prisma/issues/2444) you would need to rely on raw queries in this case.

**Reply by darrylyoung** [reply]
Hi, @ryands17! 👋 

Thanks for the reply. Good point about Prisma not supporting JSON specific filtering – I hadn't thought about that.

If I was to go ahead with that kind of solution, then, how would I go about starting? That is, if I had a basic schema like this, what would it look like if I wanted to have `description` and `notes` in the format I originally mentioned (as JSON with top-level keys as locales such as `"en"`, for example)?

```prisma
model Product {
  id          String   @id @default(cuid())
  name        String   @unique
  description String[]
  notes       String[]
}
```

I realise I need to learn a little more about this stuff – especially how to use raw queries to achieve what I want – before making a decision on whether I do this or not. Regardless, thanks for your reply.

**Reply by ryands17** [reply]
It would actually be a `Json` type like this:

```prisma
model Product {
  id          String   @id @default(cuid())
  name        String   @unique
  description Json
  notes       Json
}
```

And then you would need to query it like:

```sql
select id, name, description -> 'en', notes -> 'en' from "Product"
```

Here the language will have to be inserted dynamically via `prisma.$queryRaw` based on the chosen lang.

**Reply by darrylyoung** [reply]
Great. Thank you, as always, @ryands17.

I'll have a play around with that when I get chance. It seems like it might be the best way to go if I do decide to go for fully-translated information. Have a great weekend.

**Reply by darrylyoung** [reply]
Hi, @ryands17!

I just tried that and managed to query the correct translation via a `locale` arg that was passed to the query. That's a good start, so thanks again. I had to add `as "notes"` otherwise the result would be `?column?`, for some reason.

```js
const result = await ctx.prisma.$queryRaw<
  Product[]
>`SELECT id, name, notes -> ${locale} as "notes" from "Product"`
```

Anyway, my actual model (and the original query) is a lot more complicated than just `id`, `name`, and `notes` – there are multiple relations to other models and I'm not exactly sure how to handle that now.

When using `ctx.prisma.product.findMany`, I would automatically get back everything from a _product_, including relations, etc. Now that I'm changing a couple of fields from 'String' to 'Json', do I need to manually write out the queries for everything or is there a way I can just let Prisma do the work (as it does now) and just manually specify how I want to query the JSON for the updated `Json` fields?

I hope that makes sense. Thanks, as always, Ryan.

**Reply by ryands17** [reply]
Ohh. Actually that would make it a bit difficult as Prisma cannot currently mix raw queries with its own API. So if you need to fetch data from other tables, you might have to use the `join` keyword in `prisma.$queryRaw`. I would also suggest adding your use case to this issue: https://github.com/prisma/prisma/issues/5761 as this is something that could also be supported apart from specific JSON filtering.

**Reply by darrylyoung** [reply]
Hi, @ryands17.

Thanks for the reply. Yeah, I was worried that'd be the case. I've added a [comment](https://github.com/prisma/prisma/issues/5761#issuecomment-800937519) to #5761, explaining my situation.

To be honest, I'm not comfortable using `join` in `$queryRaw`, and writing all that stuff manually, so I'm thinking how best to approach this translation thing if I do decide to do it. Seeing as each _product_ will only have a couple of `Json` fields containing translations, what do you think about just using `findMany` as I did before but before returning the list of products in the query, filter the JSON manually by the `locale` that was passed as an argument? For example:

```js
const products = await ctx.prisma.product.findMany()

const translatedProducts = products.map((product) => ({
  ...product,
  translations: product.translations[locale],
}))

return translatedProducts
```

It's not as clean as specifying exactly which JSON you want filtered and returned from the database – and there'll be _some_ performance impact of having to do this, which I don't particularly like – but I'm not sure the overhead of returning the whole JSON and then manually filtering it before returning it would be that much of an issue. If you think it would be, though, please let me know as I'm open to suggestions.

**Reply by ryands17** [reply]
Yup that's possible and it won't be a performance overhead as you would have a limited amount of translations. It's also fine if you send the unfiltered data that is used on the frontend if you just have a couple of languages and the data is paginated.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
